### PR TITLE
Adadelta optimizer test

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
Adadelta optimizer test

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] FTML doc fix
- [x] test for adadelta

### CODE REVIEW

**Comment by vandanavk** [comment]
@mxnet-label-bot [Test, pr-awaiting-review]

File: tests/python/unittest/test_optimizer.py [hunk 1]
```diff
@@ -1002,6 +1002,83 @@ def test_adagrad():
...
+def test_adadelta():
+    opt1 = PyAdaDelta
+    opt2 = mx.optimizer.AdaDelta
+    shape = (3, 4, 5)
+    rho_options = [{'rho': 0.9}]
+    eps_options = [{}, {'epsilon': 1e-8}]
+    cg_options = [{}, {'clip_gradient': 0.4}, {'clip_gradient': 0.5}]
+    rg_options = [{}, {'rescale_grad': 0.14}, {'rescale_grad': 0.8}]
+    wd_options = [{}, {'wd': 0.0}]
+    for dtype in [np.float16, np.float32]:
```
Comment (larroy) [inline tests/python/unittest/test_optimizer.py]: itertools.product is much more legible than this staircase to heaven.

File: tests/python/unittest/test_optimizer.py [hunk 2]
```diff
@@ -1002,6 +1002,83 @@ def test_adagrad():
...
+        wd = self._get_wd(index)
+
+        grad *= self.rescale_grad
+        if self.clip_gradient is not None:
+            grad = mx.nd.clip(grad, -self.clip_gradient, self.clip_gradient)
+
+        acc_grad, acc_delta = state
+
+        acc_grad[:] = self.rho * acc_grad + (1. - self.rho) * grad ** 2
+        current_delta = mx.nd.sqrt(acc_delta + self.epsilon) / mx.nd.sqrt(acc_grad + self.epsilon) * grad
```
Comment (larroy) [inline tests/python/unittest/test_optimizer.py]: can we add parens for legibility of * / precedence?

File: tests/python/unittest/test_optimizer.py (hunk 1 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_optimizer.py]: sure I can replace it with something like this -
```
for dtype in [np.float16, np.float32]:
    for params in itertools.product(rho_options, eps_options, cg_options, rg_options, wd_options):
        kwarg = {k: v for param in params for k, v in param.items()}
```

*aside - you meant [stairway to heaven](https://www.youtube.com/watch?v=D9ioyEvdggk) right?

File: tests/python/unittest/test_optimizer.py [hunk 3]
```diff
@@ -1002,6 +1003,76 @@ def test_adagrad():
...
+    opt1 = PyAdaDelta
+    opt2 = mx.optimizer.AdaDelta
+    shape = (3, 4, 5)
+    rho_options = [{'rho': 0.9}]
+    eps_options = [{}, {'epsilon': 1e-8}]
+    cg_options = [{}, {'clip_gradient': 0.4}, {'clip_gradient': 0.5}]
+    rg_options = [{}, {'rescale_grad': 0.14}, {'rescale_grad': 0.8}]
+    wd_options = [{}, {'wd': 0.0}]
+    for dtype in [np.float16, np.float32]:
+        for params in itertools.product(rho_options, eps_options, cg_options,
```
Comment (sandeep-krishnamurthy) [inline tests/python/unittest/test_optimizer.py]: This is very clean and nice!
Thanks @larroy and @anirudhacharya

**Comment by larroy** [comment]
yes, good job.

**Pull request merged**